docs/topics/request-queues.rst

```diff
@@ -0,0 +1,160 @@
+queue that Scrapy uses. Its default value is :class:`ScrapyPriorityQueue`.
+
+Scrapy provides the following priority queue implementations:
+
+.. autoclass:: DownloaderAwarePriorityQueue
+
+.. autoclass:: ScrapyPriorityQueue
+
+
+.. setting:: SCHEDULER_DISK_QUEUE
+.. setting:: SCHEDULER_MEMORY_QUEUE
+.. _downstream-queues:
+
+Downstream queues
+=================
+
+.. currentmodule:: scrapy.squeues
+
+The built-in :ref:`priority queues <priority-queues>` use queues internally to
+store pending requests. These internal queues are called *downstream queues*.
+
+The organization method used by the downstream queues, usually either
+FIFO or LIFO, affects the order in which requests are sent for requests
+that the :ref:`priority queue <priority-queues>` considers to have the same
+priority.
+
+The built-in :ref:`priority queues <priority-queues>` allow you to choose the
+type of downstream queue they use through a setting. Which setting, however,
+depends on the value of the :setting:`JOBDIR` setting (see :ref:`topics-jobs`):
+
+-   If :setting:`JOBDIR` is not set (default), the memory downstream queue type
+    defined in the :setting:`SCHEDULER_MEMORY_QUEUE` setting is used, which
+    defaults to :class:`LifoMemoryQueue`.
+
+    Scrapy provides the following memory downstream queue classes:
+
+    .. autoclass:: FifoMemoryQueue
+
+    .. autoclass:: LifoMemoryQueue
+
+-   If :setting:`JOBDIR` is set, the disk downstream queue from the
+    :setting:`SCHEDULER_DISK_QUEUE` setting is used instead, which defaults to
+    :class:`PickleLifoDiskQueue`.
+
+    In addition to request order, which disk downstream queue you use can
+    affect request serialization speed (CPU usage) and size (disk usage). See
+    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
+    more information.
+
+    When request serialization fails, Scrapy falls back to using a memory
+    downstream queue.
+
+    Scrapy provides the following disk downstream queue classes:
+
+    .. autoclass:: MarshalFifoDiskQueue
+
+    .. autoclass:: MarshalLifoDiskQueue
+
+    .. autoclass:: PickleFifoDiskQueue
+
+    .. autoclass:: PickleLifoDiskQueue
+
+
+Writing your own disk downstream queue class
+--------------------------------------------
+
+If you want to define and use your own disk downstream queue implementation,
+it has to conform to the following interface:
+
+.. class:: MyExternalQueue
+
+   .. classmethod:: __init__(cls, crawler: scrapy.crawler.Crawler, key: str, state)
+
+      Initializes an instance of this disk downstream queue class.
+
+      *key* is the unique ID of the disk downstream queue instance. It may be
+      used, for example, to create a unique file or folder to store the queue
+      content.
+
+      *state* is a variable containing state of queue. It should be serializeable
+      and deserializeble by json module. You need to come up with proper type
+      and structure to completely restore an internal state of your queue.
```

Comment:
Thanks, I think I now understand how this works from just the documentation.

What would you think about a bit of rephrasing?

```suggestion
      *state* is the data returned by the :method:`close` method of this queue 
      during a prior execution, and should be used to restore the queue to that 
      prior state. If there is no prior execution, its value is ``None``.
```

With a reference to `close` here now, I think it makes sense to keep the rest 
of the information in the documentation of the `close` method that is 
responsible for creating the state data.